### ISSUE TITLE
Impossible to search by node-label or relationship-type

### ISSUE THREAD

**Issue by davidrapin** [graphaware/neo4j-to-elasticsearch#11]
Right now, the node-label/relationship-type is stored in the `_type` (not analyzed). In the future I understand that you are considering creating one ES index per node-label/relationship-type.

In both cases, it is impossible to search for nodes/relationships by node-label/relationship-type.

For example, if some of my nodes have the "foo bar" label, if would like to find them by searching for `"foo"` or `"bar"` or `"foo bar"`.
```bash
curl -XPUT -d '{"prop1":"abc"}' http://localhost:9200/myindex_nodes/foo%20bar/1
{"_index": "myindex_nodes", "_type": "foo bar", "_id": "1", "_version": 1, "created": true}

curl -XGET http://localhost:9200/myindex_nodes/_search?q=foo
{"hits": {"total": 0}}
```

A solution would be to store the node-label(s)/relationship-type in a special field (`neo4j_label`), e.g.:
```bash
curl -XPUT -d '{"neo4j_labels":["foo bar"], "prop1":"abc"}' http://localhost:9200/myindex_nodes/foo%20bar/2
{"_index":"myindex_nodes", "_type":"foo bar", "_id":"2", "_version":1, "created": true}

curl -XGET http://localhost:9200/myindex_nodes/_search?q=foo
{
  "hits": {
    "total":1, 
    "hits": [
      {
        "_index": "my_index",
        "_type": "foo bar",
        "_id": "2",
        "_score": 1.1931472,
        "_source": {
          "neo4j_labels": ["foo bar"],
          "prop1": "abc"
        }
      }
    }
  ]
}
```

**Comment by ikwattro** [comment]
Definitely !!!

> For example, if some of my nodes have the "foo bar" label, if would like to find them by searching for "foo" or "bar" or "foo bar".

You can specify mulitple indices in your search url.

But anyway, a field representing the type (should be labels or relationship type) is needed.

**Comment by alenegro81** [comment]
this is a good catch.

**Comment by ikwattro** [comment]
What we propose is the following : 

Nodes and relationships will be indexed in their own `index`. If a node has multiple labels, it will be indexed in each `labeled index`.

There will be 2 fields : 

* 1 representing a `label` or `relationship type` -> let's call it `internal_type` for now
* 1 representing `all labels` -> lets's call it `internal_all` types for now

In that way you can query per label, per rel type, per "all labels" and apply also filters/function_scores on every of them.

Would that suit your needs @davidrapin  ?

**Comment by davidrapin** [comment]
@ikwattro Looking good. I don't see the use case for `internal_type`  (with only the label from the current index?).
Since labeled indices are used (`myprefix_nodes_labelFoo`, `myprefix_nodes_labelBar` etc.), filtering on a specific label can be achieved by querying that labeled index explicitly (e.g. `GET http://localhost:9200/myprefix_nodes_labelFoo` ).

For the field name, I would advise prefixing with underscores and a unique string to minimize the collision risk with actual node/relationship fields (something like ` __n2e_types`).

**Comment by ikwattro** [comment]
Use case is more for running function scores queries based on this value

**Comment by davidrapin** [comment]
@ikwattro You can avoid the side-effect of lower score in array with more values by disabled `norms` for that field in the mapping. See http://stackoverflow.com/questions/30346557/elastic-search-array-score

**Comment by ikwattro** [comment]
I was more thinking of boosting based on the tf/idf/bm25 score of this field, thinking if it would be worth though

**Comment by davidrapin** [comment]
I think boosting based on tf/idf would work as expected for a single value if the array is mapped with `norms: {enabled: false}`

**Comment by ikwattro** [comment]
Fixed by #21
